mammoth/utils/optimizers.py

```diff
@@ -181,243 +138,255 @@ def linear_warmup_decay(step, warmup_steps, rate, train_steps):
+        }
+        for name, optimizer in self.suboptimizers.items():
+            if name in trained_components:
+                # logger.warning(f'Stepping {name}')   # DEBUG
+                if self.grad_scaler is not None:
+                    self.grad_scaler.unscale_(optimizer)
+                optimizer.step()
 
-    def learning_rate(self):
-        """Returns the current learning rate."""
-        if self._learning_rate_decay_fn is None:
-            return self._learning_rate
-        scale = self._learning_rate_decay_fn(self._decay_step)
-        return scale * self._learning_rate
+        if self.grad_scaler is not None:
+            # Updates the scale for next iteration.
+            self.grad_scaler.update()
+
+    def report_steps(self):
+        result = []
+        for name, optimizer in self.suboptimizers.items():
+            count = optimizer.training_step
+            lr = optimizer.learning_rate()
+            result.append(f'Optimizer "{name}" has been stepped {count} times and has LR {lr}')
+        return result
 
     def state_dict(self):
+        """Returns the state dictionary"""
         return {
-            'training_step': self._training_step,
-            'decay_step': self._decay_step,
-            'optimizer': self._optimizer.state_dict(),
+            'optimizers': {k: v.state_dict() for k, v in self.suboptimizers.items()},
+            'steps': self._steps,
         }
 
     def load_state_dict(self, state_dict):
-        self._training_step = state_dict['training_step']
-        # State can be partially restored.
-        if 'decay_step' in state_dict:
-            self._decay_step = state_dict['decay_step']
-        if 'optimizer' in state_dict:
-            self._optimizer.load_state_dict(state_dict['optimizer'])
+        """Loads the optimizer from the state dictionary"""
 
-    def zero_grad(self):
-        """Zero the gradients of optimized parameters."""
-        self._optimizer.zero_grad()
+        # do not load any optimizer state if one component is missing
+        do_load = True
+        for k in state_dict["optimizers"].keys():
+            if k not in self.suboptimizers.keys():
+                do_load = False
 
-    def backward(self, loss):
-        """Wrapper for backward pass. Some optimizer requires ownership of the
-        backward pass."""
-        if self.amp:
-            self._scaler.scale(loss).backward()
-        elif self._fp16 == "legacy":
-            kwargs = {}
-            if "update_master_grads" in fn_args(self._optimizer.backward):
-                kwargs["update_master_grads"] = True
-            self._optimizer.backward(loss, **kwargs)
+        if do_load is True:
+            for k in state_dict["optimizers"].keys():
+                self.suboptimizers[k].load_state_dict(state_dict["optimizers"][k])
         else:
-            loss.backward()
+            logger.info("Some components do not match. Do not load optimizer from checkpoint.")
```

Comment:
can we be explicit with the components extra/not found? perhaps in a debug message (to be fixed, ik)

Reply:
I actually didn't add this line, the diff is just wonky.
Seems to be added by @josephattieh in commit e3d8a64f8